src/ydata_synthetic/synthesizers/regular/cgan/model.py

```diff
@@ -65,23 +81,30 @@ def get_data_batch(self, train, batch_size, seed=0):
         stop_i = start_i + batch_size
         shuffle_seed = (batch_size * seed) // len(train)
         np.random.seed(shuffle_seed)
-        train_ix = np.random.choice(list(train.index), replace=False, size=len(train))  # wasteful to shuffle every time
+        train_ix = np.random.choice(train.shape[0], replace=False, size=len(train))  # wasteful to shuffle every time
         train_ix = list(train_ix) + list(train_ix)  # duplicate to cover ranges past the end of the set
-        x = train.loc[train_ix[start_i: stop_i]].values
-        return np.reshape(x, (batch_size, -1))
+        return train[train_ix[start_i: stop_i]]
 
-    def train(self, data: Union[DataFrame, array],
-              label:str,
-              train_arguments:TrainParameters):
+    def train(self, data: Union[DataFrame, array], label:str, train_arguments: TrainParameters, num_cols: List[str],
+              cat_cols: List[str], preprocess: bool = True):
         """
         Args:
             data: A pandas DataFrame or a Numpy array with the data to be synthesized
             label: The name of the column to be used as a label and condition for the training
-            train_arguments: Gan training arguments.
-        Returns:
-            A CGAN model fitted to the provided data
+            train_arguments: GAN training arguments.
+            num_cols: List of columns of the data object to be handled as numerical
+            cat_cols: List of columns of the data object to be handled as categorical
+            preprocess: If True preprocess the data before using in train session
         """
-        iterations = int(abs(data.shape[0] / self.batch_size) + 1)
+        super().train(data, num_cols, cat_cols, preprocess)
+
+        processed_data = self.processor.transform(data)
+        self.data_dim = processed_data.shape[1] - len(self.processor.cat_pipeline.get_feature_names_out())
```

Comment:
From the data dim, why are we removing cat columns from the preprocessor dimension. 

I should remind that the class column shouldn't go through the preprocessing.